ckanext/search/filters.py

```diff
@@ -4,51 +4,189 @@
 
 """
 
+from typing import NamedTuple, Any, Dict, List, Union, Optional
+
 from ckan.plugins.toolkit import Invalid
 
 
-TOP_OPERATORS = ["$or", "$and"]
+class FilterOp(NamedTuple):
+    op: str
+    field: Optional[str]
+    value: Any
+
+    def __repr__(self) -> str:
+        if (
+            isinstance(self.value, list)
+            and self.value
+            and isinstance(self.value[0], FilterOp)
+        ):
+            child_reprs = []
+            for child in self.value:
+                child_lines = str(child).split("\n")
+                indented_lines = ["    " + line for line in child_lines]
+                child_reprs.append("\n".join(indented_lines))
+
+            value_str = "[\n" + ",\n".join(child_reprs) + "\n]"
+        else:
+            value_str = repr(self.value)
+
+        return (
+            f"FilterOp(op={repr(self.op)}, field={repr(self.field)}, value={value_str})"
+        )
+
+
+FILTER_OPERATORS = ["$or", "$and"]
+
+
+def _is_dict_or_list_of_dicts(value: Any) -> bool:
+    return isinstance(value, dict) or _is_list_of_dicts(value)
+
+
+def _is_list_of_dicts(value: Any) -> bool:
+    return isinstance(value, list) and all(isinstance(item, dict) for item in value)
+
+
+def _check_filter_operator(key: str, value: Any) -> None:
+    if key not in FILTER_OPERATORS:
+        raise Invalid(f"Unknown operators (must be one of $or, $and): {key}")
+
+    # Filter operator values must be lists of dicts
+    if not _is_list_of_dicts(value):
+        raise Invalid("Filter operations must be defined as a list of dicts")
+
+
+def _process_filter_operator_members(value: Any) -> List[Dict[str, Any]]:
+
+    if not isinstance(value, list):
+        raise Invalid("Filter operations must contain lists of filters")
+
+    out: List[Dict[str, Any]] = []
+    for item in value:
+        if not isinstance(item, dict):
+            raise Invalid("Filter operation members must be dictionaries")
+
+        if len(item.keys()) > 1:
+            raise Invalid("Field operation can only have one key")
+
+        key = list(item.keys())[0]
+
+        if key.startswith("$") and not key.startswith("$$"):
+            _check_filter_operator(key, item[key])
+            item[key] = _process_filter_operator_members(item[key])
+
+            out.append(item)
+        else:
+            result = _process_field_operator(key, item[key])
+            if "$or" in result or "$and" in result:
+                out.append(result)
+            else:
+                item[key] = result
+
+                out.append(item)
+
+    return out
+
+
+def _to_filter_op(filters_dict: dict) -> Optional[FilterOp]:
+
+    out = []
+    for k, v in filters_dict.items():
+        if k in FILTER_OPERATORS:
```

Comment:
for consistency with above (and to allow custom filter operators in the future)
```suggestion
        if k.startswith("$") and not k.startswith("$$"):
```